### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by vanhauser-thc**
we should merge to master again ... it has been so much change

### CODE REVIEW

File: src/afl-fuzz-one.c [hunk 1]
```diff
@@ -1974,8 +1996,9 @@ u8 fuzz_one_original(afl_state_t *afl) {
             memcpy(new_buf + clone_to + clone_len, out_buf + clone_to,
                    temp_len - clone_to);
 
-            ck_free(out_buf);
+            swap_bufs(BUF_PARAMS(out), BUF_PARAMS(out_scratch));
```
[the remark this answers is not in the document]
Comment (domenukk) [reply src/afl-fuzz-one.c:1999]: it should: `out_buf` and `out_scratch_buf` are in `afl_state_t`. (The `BUF_PARAMS` covers this up, maybe that could be refactored)
Each thread will (probably) have its own state.

File: src/afl-fuzz-mutators.c [hunk 2]
```diff
@@ -225,12 +242,7 @@ u8 trim_case_custom(afl_state_t *afl, struct queue_entry *q, u8 *in_buf) {
     fault = run_target(afl, afl->fsrv.exec_tmout);
     ++afl->trim_execs;
 
-    if (afl->stop_soon || fault == FAULT_ERROR) {
-
-      ck_free(retbuf);
-      goto abort_trimming;
-
-    }
+    if (afl->stop_soon || fault == FAULT_ERROR) { goto abort_trimming; }
```
[the remark this answers is not in the document]
Comment (domenukk) [reply src/afl-fuzz-mutators.c:245]: Indeed. We have slightly different documentations at 3 places now :/
Changed it in this place.

File: src/afl-fuzz-init.c [hunk 3]
```diff
@@ -384,8 +391,7 @@ void read_testcases(afl_state_t *afl) {
 
     if (!S_ISREG(st.st_mode) || !st.st_size || strstr(fn2, "/README.txt")) {
 
-      ck_free(fn2);
-      ck_free(dfn);
+      free(fn2);
```
[the remark this answers is not in the document]
Comment (domenukk) [reply src/afl-fuzz-init.c:394]: Yes, `ck_free` (probably we should rename `alloc_printf` to `ck_alloc_printf` against this confusion.
We cannot use a non-malloced buffer here, as the buffer is stored in the queue after the function's lifetime.

File: Makefile [hunk 4]
```diff
@@ -303,21 +304,28 @@ afl-tmin: src/afl-tmin.c src/afl-common.o src/afl-sharedmem.o src/afl-forkserver
...
 
 # document all mutations and only do one run (use with only one input file!)
 document: $(COMM_HDR) include/afl-fuzz.h $(AFL_FUZZ_FILES) src/afl-common.o src/afl-sharedmem.o src/afl-forkserver.o | test_x86
 	$(CC) -D_AFL_DOCUMENT_MUTATIONS $(CFLAGS) $(CFLAGS_FLTO) $(AFL_FUZZ_FILES) src/afl-common.o src/afl-sharedmem.o src/afl-forkserver.o -o afl-fuzz-document $(PYFLAGS) $(LDFLAGS)
 
+test/unittests/unit_maybe_alloc.o : $(COMM_HDR) include/alloc-inl.h test/unittests/unit_maybe_alloc.c $(AFL_FUZZ_FILES)
+	$(CC) $(CFLAGS) $(CFLAGS_FLTO) -c test/unittests/unit_maybe_alloc.c -o test/unittests/unit_maybe_alloc.o
+
+unit_maybe_alloc: test/unittests/unit_maybe_alloc.o
+	$(CC) $(CFLAGS) -Wl,--wrap=exit -Wl,--wrap=printf $(LDFLAGS) test/unittests/unit_maybe_alloc.o -o test/unittests/unit_maybe_alloc -ldl -lcmocka
```
[the remark this answers is not in the document]
Comment (domenukk) [reply Makefile:319]: I don't find any documentation but a cmocka package already existed on ubuntu prior to 16.04 and it states "It only requires the standard C library, works on a range of computing platforms (including embedded) and with different compilers."
The log you sent earlier indicates wrapping worked but something else went wrong in the linker. We should investigate.

File: src/afl-fuzz-run.c [hunk 5]
```diff
@@ -214,16 +205,22 @@ void write_to_testcase(afl_state_t *afl, void *mem, u32 len) {
```
[the remark this answers is not in the document]
Comment (domenukk) [reply src/afl-fuzz-run.c:205]: Nice! :+1:

File: include/afl-fuzz.h [hunk 6]
```diff
@@ -928,7 +965,7 @@ static inline u32 get_rand_seed(afl_state_t *afl) {
 /* Find first power of two greater or equal to val (assuming val under
    2^63). */
 
-static u64 next_p2(u64 val) {
+static inline u64 next_p2(u64 val) {
```
[the remark this answers is not in the document]
Comment (domenukk) [reply include/afl-fuzz.h:968]: Done.

File: examples/README.md [hunk 7]
```diff
@@ -3,8 +3,6 @@
 Here's a quick overview of the stuff you can find in this directory:
 
   - custom_mutstors      - An example custom mutator
-  
-  - python_mutators      - Python mutators examples
```
[the remark this answers is not in the document]
Comment (domenukk) [reply examples/README.md:7]: I've changed it to `example custom mutators in python an c`

File: dictionaries/regexp.dict [hunk 8]
```diff
@@ -238,366 +238,19 @@
 "\\p{Nd}"
 "\\P{Any}"
 "\\p{Changes_When_NFKC_Casefolded}"
-"L~"
```
[the remark this answers is not in the document]
Comment (domenukk) [reply dictionaries/regexp.dict:241]: His PR removed a few hundred redundant lines.

**Comment by domenukk** [comment]
Worked in many of @hexcoder-'s good comments.
Somebody else may resolve the open questions.

File: src/afl-forkserver.c [hunk 9]
```diff
@@ -318,24 +315,15 @@ void afl_fsrv_start(afl_forkserver_t *fsrv, char **argv) {
...
-    fd_set readfds;
-
-    FD_ZERO(&readfds);
-    FD_SET(fsrv->fsrv_st_fd, &readfds);
-    timeout.tv_sec = ((fsrv->exec_tmout * FORK_WAIT_MULT) / 1000);
-    timeout.tv_usec = ((fsrv->exec_tmout * FORK_WAIT_MULT) % 1000) * 1000;
-
-    int sret = select(fsrv->fsrv_st_fd + 1, &readfds, NULL, NULL, &timeout);
+    rlen = 4;
+    u32 time = read_timed(fsrv->fsrv_st_fd, &status, rlen,
```
[the remark this answers is not in the document]
Comment (domenukk) [reply src/afl-forkserver.c:319]: For me it does. If it doesn't please let me know, in that case we'll have to rewrite read_timed.

File: dictionaries/regexp.dict (hunk 8 above)
[the remark this answers is not in the document]
Comment (domenukk) [reply dictionaries/regexp.dict:241]: Ah you mean keep them? Yeah, I don't know either...

File: docs/Changelog.md [hunk 10]
```diff
@@ -16,12 +16,15 @@ sending a mail to <afl-users+subscribe@googlegroups.com>.
...
   - all:
-    - big code changes to make afl-fuzz thread-safe so afl-fuzz can spawn 
+    - big code changes to make afl-fuzz thread-safe so afl-fuzz can spawn
       multiple fuzzing threads in the future or even become a library
     - afl basic tools now report on the environment variables picked up
     - more tools get environment variable usage info in the help output
     - force all output to stdout (some OK/SAY/WARN messages were sent to
       stdout, some to stderr)
+    - uninstrumented mode uses an internal forkserver ("fauxserver")
+    - now builds with `-D_FORTIFY_SOURCE=2`
```
[the remark this answers is not in the document]
Comment (vanhauser-thc) [reply docs/Changelog.md:26]: We compile with -O3 so we are fine (and compiling afl without optimization would be bad :) ) - however if FORITFY_SOURCE really improves performance we still need to test

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (vanhauser-thc) [reply Makefile:319]: How about we only do unit tests in test.sh if we have a specific version or higher?

[the remark this answers is not in the document]
Comment (domenukk) [reply Makefile:319]: Yes I think this is the best solution going forward. For now it works, but once we need more features, we should fix it to v 1.0 or higher

**Pull request merged**